globals/matchmaking_utils.go

```diff
@@ -9,24 +10,53 @@ import (
 	"github.com/PretendoNetwork/nex-go/v2"
 	"github.com/PretendoNetwork/nex-go/v2/constants"
 	"github.com/PretendoNetwork/nex-go/v2/types"
+	"github.com/PretendoNetwork/nex-protocols-go/v2/globals"
 	match_making "github.com/PretendoNetwork/nex-protocols-go/v2/match-making"
 	match_making_types "github.com/PretendoNetwork/nex-protocols-go/v2/match-making/types"
 	notifications "github.com/PretendoNetwork/nex-protocols-go/v2/notifications"
 	notifications_types "github.com/PretendoNetwork/nex-protocols-go/v2/notifications/types"
 	"golang.org/x/exp/slices"
 )
 
+var sessions map[uint32]*CommonMatchmakeSession
+var sessionsMutex = sync.RWMutex{}
+
+var SessionManagementDebugLog = false
+
+func MakeSessions() {
+	sessions = make(map[uint32]*CommonMatchmakeSession)
+}
+
+func GetSession(gatheringID uint32) (*CommonMatchmakeSession, bool) {
+	sessionsMutex.RLock()
+	defer sessionsMutex.RUnlock()
+
+	ret, ok := sessions[gatheringID]
+	return ret, ok
+}
+
+func EachSession(callback func(index uint32, value *CommonMatchmakeSession) bool) bool {
+	sessionsMutex.RLock()
+	defer sessionsMutex.RUnlock()
+
+	for i, value := range sessions {
+		if callback(i, value) {
+			return true
+		}
+	}
+
+	return false
+}
+
 // GetAvailableGatheringID returns a gathering ID which doesn't belong to any session
 // Returns 0 if no IDs are available (math.MaxUint32 has been reached)
 func GetAvailableGatheringID() uint32 {
 	return CurrentGatheringID.Next()
 }
 
-// FindOtherConnectionID searches a connection ID on the session that isn't the given one
-// Returns 0 if no connection ID could be found
-func FindOtherConnectionID(excludedConnectionID uint32, gatheringID uint32) uint32 {
+func findOtherConnectionIDImpl(excludedConnectionID uint32, gatheringID uint32) uint32 {
```

Comment:
I'm not a *huge* fan of the way these `Impl` functions are named/handled, though I understand why the implementation is separate from the exported function

This `Impl` convention could be refactored out by moving all of this out of the global scope and moving it to a struct, and having the functions currently named with `Impl` as just unexported methods

Something like

```go
package main

import "fmt"

type sessionsManager struct{}

func (sm *sessionsManager) FindOtherConnectionID() {
	fmt.Println("Exported")

	sm.findOtherConnectionID()
}

func (sm *sessionsManager) findOtherConnectionID() {
	fmt.Println("Unexported")
}

func main() {
	sm := &sessionsManager{}

	sm.FindOtherConnectionID()
	// Prints both Exported and Unexported
}
```

@DaniElectra thoughts?

Reply:
Waiting for DaniElectra response, it's also possible to not export the functions by just making them lowercase. But it may be better to have the sessions maps in the struct instead of being something global.

Reply:
> it's also possible to not export the functions by just making them lowercase

That's already how it's implemented, both in your PR and my suggestion? My comment was about exported vs unexported, it was on the naming conventions of the methods

It's clear you want to export a function which calls an implementation, without exposing the implementation itself. This is fine, I'm just not a fan of the way the functions themselves are named currently. So my suggestion was to bring them into a "manager" style struct, renaming the `Impl` functions to just have the same name as the exported one but be unexported

Though now that I think about it, why *is* the implementation functions separate now? I see elsewhere you call the `Impl` functions directly, but since both the exported and `Impl` functions have the same signatures why not just call the exported function directly and not have this separation?

> But it may be better to have the sessions maps in the struct instead of being something global.

My example was minimal to try and get the point across but yes ideally in this example everything to do with sessions would be contained in the manager (the lock, the map, etc)

Reply:
The impl is needed to be able to call functions without locking the sessions lock again. I will move everything to a struct after I figure out the stuck gathering issues.